Update readme for client and other packages

krisbitney opened the pull request:
This is a draft PR for the purpose of preparing readme documents for the 0.10.0 release. The readme docs will be used for reference documentation on our docs site.

Steps to update a typical readme with reference documentation:
1. If a readme does not exist, create one with an introduction, some usage examples, etc.
2. Add doc comments to exported methods and types.
3. Add `doc-snippets` tokens around the comments and types or method signatures.
4. Use the `doc-snippets` CLI tool to inject the snippets into the readme doc.

You can copy the `doc-snippets` setup and configuration from repos where this has been done.

The aforementioned steps won't make sense for every package. Feel free to adapt as necessary!

TODO:
- [x] @polywrap/client-js
- [x] @polywrap/core-client-js
- [x] @polywrap/client-config-builder-js
- [x] @polywrap/core-js
- [x] polywrap (CLI)
- [x] @polywrap/test-env-js
- [ ] @polywrap/uri-resolvers-js
- [ ] @polywrap/uri-resolver-extensions-js
- [ ] @polywrap/package-validation
- [ ] @polywrap/wasm-js
- [ ] @polywrap/plugin-js
- [ ] @polywrap/tracing-js
- [ ] @polywrap/result
- [ ] @polywrap/msgpack-js
- [ ] @polywrap/wrap-manifest-types-js
- [ ] @polywrap/polywrap-manifest-types-js
- [ ] @polywrap/logging-js
- [ ] @polywrap/asyncify-js
- [ ] @polywrap/os-js

krisbitney committed:
updated readme for client and core-client

pileks commented on packages/js/client/README.md:
```diff
@@ -70,15 +82,14 @@ const config = {
+  resolvers: [
+    new RecursiveResolver(
+      new PackageToWrapperCacheResolver(wrapperCache, [
+        new ExtendableUriResolver(),
```
Do we have any docs that explain the various resolvers, at least the ones that come with Polywrap by default?

This might help developers set up their own resolution logic.

I'm thinking we could add this to `packages/js/uri-resolver-extensions/README.md` and reference it here? @nerfZael thoughts?

krisbitney replied:
Imo that's a good idea. I can work on it soon. I was also thinking to write more detailed reference documentation (in the readme) for the client config builder since it's becoming so important.

It would be cool to create good reference docs (as readmes) for wasm-js and plugin-js, and really every package.

pileks replied:
Ideally we'd have the package creators also write up documentation, or at least a basic outline of it (with functional details), and have someone else (e.g. me, you, any volunteer) fix it up to spec.

krisbitney replied:
yes, 100%

pileks reviewed (dismissed):
Overall LGTM! Comment is non-blocking.

cbrzn commented on packages/js/core-client/README.md:
````diff
@@ -6,3 +6,382 @@
+
+```ts
+await client.invoke({
+  uri: "ens/rinkeby/helloworld.dev.polywrap.eth",
````
```suggestion
  uri: "ens/helloworld.dev.polywrap.eth",
```

cbrzn replied:
also: i would also suggest this in the `client/README.md` but gh doesn't allow me

cbrzn commented on packages/js/client/README.md:
```diff
@@ -7,6 +7,8 @@
 <br/>
 The Polywrap JavaScript client invokes functions of wrappers and plugins. It's designed to run in any environment that can execute JavaScript (think websites, node scripts, etc.). It has TypeScript support.
```
should this paragraph be moved to `core-client`? the reasoning is that this package doesn't take the responsibility of the invocation (or any), it just exists to have an ez way to config the client, right?

cbrzn commented on packages/js/client/README.md:
```diff
@@ -45,11 +47,21 @@ const config = {
   ],
-  // declare and configure plugin wrappers
-  plugins: [
+  // add embedded Wasm wrappers
```
in line `43` of this file, there's `// redirect queries from one uri to another` - I would recommend to change it to `// redirect invocation from one uri to another`

it doesn't allow me to add the comment in that line because it wasn't modified lol

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

krisbitney committed:
removed "rinkeby" from client readme invocation example; moved description of polywrap client in readme to core client; replaced "queries" word with "invocations"

krisbitney committed:
removed "rinkeby" from core client readme invocation example

krisbitney committed:
added TSDoc comments to client config builder and added reference section to client config builder README.md

krisbitney committed:
added client configs to client, core client, and config builder readmes

krisbitney committed:
adjusted client config comments to more closely match formatting used elsewhere for interface types

krisbitney committed:
added TSDoc comments for CoreClient.ts

krisbitney committed:
replaced @polywrap/client-js readme reference content with doc snippet markers

krisbitney committed:
using doc-snippets package to build readme in @polywrap/client-js

krisbitney committed:
using doc-snippets package to build readme in @polywrap/core-client-js

krisbitney committed:
using doc-snippets package to build readme in @polywrap/client-config-builder-js

krisbitney committed:
using doc-snippets package to build readme in @polywrap/core-js

krisbitney committed:
Added TSDoc comments throughout @polywrap/core-js and ensured they are pulled into readme by doc-snippets

krisbitney committed:
lint

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

krisbitney committed:
updated doc-snippets version to 0.4.0-pre.4 to use inline snippets

pileks commented:
Should we consider merging this and opening documentation issues for the remaining packages?

krisbitney committed:
added doc comments and readme updates to test-env-js

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

# Conflicts:
#	packages/js/client-config-builder/src/bundles/getDefaultConfig.ts
#	packages/js/core-client/src/PolywrapCoreClient.ts
#	packages/js/core/src/interfaces/uri-resolver.ts
#	packages/js/test-env/src/index.ts

krisbitney committed:
merge conflict fix

krisbitney commented:
> Should we consider merging this and opening documentation issues for the remaining packages?

Yeah, I think that's a good idea. It will prevent later work fixing merge conflicts.

krisbitney committed:
fixed typo made when resolving merge conflicts

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

# Conflicts:
#	packages/js/client-config-builder/package.json
#	packages/js/client/package.json
#	packages/js/core-client/package.json

dOrgJelli commented on packages/js/client-config-builder/README.md:
````diff
@@ -42,15 +42,470 @@ const config = builder.build();
+
+### Constructor
+```ts
+   * Instantiate a ClientConfigBuilder
````
This section doesn't appear to be rendering correctly, it's missing the starting `/*`

krisbitney replied:
fixed!

dOrgJelli commented on packages/js/client-config-builder/README.md:
````diff
@@ -42,15 +42,470 @@ const config = builder.build();
 ```
 
-## Methods
+## Types
````
Above this line, in the introduction section, I think it could use some work. A few thoughts:
- I don't think "DSL" is properly used here.
- Ideally all code snippets should be from-source so that they remain up-to-date, so this one that's not a snippet worries me as it can become out-of-date.
- I think should start with introducing the main concepts a user must understand to get started. We can introduce the intended usage pattern of `init -> configure -> build`, and we can point to sub-sections for each. "Initialization" shows the constructor signature. "Configure" shows the different "add..." methods. etc
- Lastly, in this intro section, having an e2e example snippet (taken from a buildable source file) would be great to have.

So in summary:
- Brief Description
- Intro Concepts (linking to reference sections further down)
- E2E Example Snippet

Apologies if this is overly verbose :P

krisbitney replied:
i reworked this intro and used snippets as suggested in your other comment

dOrgJelli commented on packages/js/client-config-builder/README.md:
````diff
@@ -42,15 +42,470 @@ const config = builder.build();
+}
+```
+
+## ClientConfigBuilder
````
Can we group the `ClientConfigBuilder` and `ClientConfig` sections underneath a `Reference` section?

krisbitney replied:
done!

dOrgJelli commented on packages/js/client-config-builder/package.json:
```diff
@@ -16,7 +16,8 @@
     "test:ci": "jest --passWithNoTests --runInBand --verbose",
-    "test:watch": "jest --watch --passWithNoTests --verbose"
+    "test:watch": "jest --watch --passWithNoTests --verbose",
+    "build:readme": "yarn doc-snippets combine"
```
I worry that this "build:readme" can be easily forgotten about.

I think for inspiration we can  look at how we build the CLI package. We have a `build` command which does a lot of different build steps, and then there is `build:fast` which only builds the source files.

Maybe a similar approach here would make sense?

krisbitney replied:
done!

dOrgJelli commented on packages/js/client-config-builder/package.json:
```diff
@@ -45,5 +47,41 @@
+        "inline": false
+      },
+      {
+        "pattern": "/* $: {SNIPPET_NAME} */",
```
Why do we have `{SNIPPET_NAME}` here, but not in the above start token?

dOrgJelli replied:
Oh nvm, I think I figured it out, it's because this one is "inline" so it can't know where to extract the snippet name from like it can in the pattern above. Is that correct?

krisbitney replied:
right, that's correct. The `{SNIPPET_NAME}` helps define a regex pattern.

dOrgJelli commented on packages/js/client/readme/README.md:
````diff
@@ -0,0 +1,118 @@
+### Invoke a wrapper
+
+```ts
+await client.invoke({
````
We should convert this to a code snippet so we make sure it runs. Maybe extract it from the tests folder or something.

dOrgJelli replied:
Also it seems like we have this problem with the other READMEs I'm seeing here, all the example code snippets aren't being extracted from source. Might I suggest us adding an `examples/` folder next to the `src/` folder where we keep these examples? They would be built so we can make sure they are correct, but their artifacts will not make it into the `build/` folder so they do not get published with the built `src/` files.

dOrgJelli replied:
Another option is to create a `src/__snippets__` folder, that lives alongside the `src/__tests__` folder.

krisbitney replied:
I added `examples/` folders and set up snippets in `client`, `core-client`, `client-config-builder`, and `test-env`.

dOrgJelli reviewed (changes requested):
This all looks great! Just a few improvements I think are needed before merging.

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

krisbitney committed:
fixed typo in client config builder constructor snippet extraction; added "build:readme" to build scripts and added "build:fast" options; added "Reference" section headers to readmes

krisbitney committed:
reworked intro in client config builder readme using buildable snippets

krisbitney committed:
reworked doc-snippets config for client config builder to properly extract snippets from examples

krisbitney committed:
added snippets to client readme intro

krisbitney committed:
added snippets to core client readme intro

krisbitney committed:
added snippets to test-env-js readme intro

krisbitney commented:
@dOrgJelli Your suggestions were really good. Thanks for looking at this! 

I think I addressed all of your comments: I updated the client config builder readme. All code examples now use snippets that build with `yarn build`. We can use `yarn build:fast` to just build the source code. 

I think it's in much better shape and ready for a new review!

dOrgJelli reviewed (dismissed):
:clap: :fire: :heavy_check_mark:

dOrgJelli committed:
Merge branch 'origin-dev' into update-client-readme

krisbitney committed:
Merge remote-tracking branch 'origin/origin-dev' into update-client-readme

# Conflicts:
#	packages/js/client-config-builder/src/ClientConfigBuilder.ts
#	packages/js/core/src/uri-resolution/UriResolutionContext.ts
#	packages/js/test-env/src/index.ts
#	yarn.lock

krisbitney committed:
readme updates
